designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
```

Comment:
This is inconsistent with our current experience. Do you propose doing this for all other explorer nodes?

Reply:
> Do you propose doing this for all other explorer nodes?

I think it makes sense for discoverability. Hiding the nodes just hides the problem that users would encounter when they are using all of the services. So we should solve that problem instead of hiding it.

Reply:
The current experience is to have a child node that says "[No x found]".
I don't think VS Code currently has a construct to "grey out" nodes, but you can control the node's expansion state (collapsed, expanded, or not present), and if you know there are 0 buckets, you could set the expander to be not present.

Reply:
If we *are* going to show the counts / grey out:
- We need to be consistent and show the counts for every resource type (again this could potentially get expensive)
- We need to make sure we're not locking the UI thread / stopping them from interacting with other nodes in the explorer while we do these pre-emptive fetches

Reply:
> the expander is _always_ present.

That's an anxiety UX. It _is_ a common pattern because it's easier for developers to implement, but I wouldn't elevate to the level of a _convention_.

> We need to be consistent and show the counts for every resource type (again this could potentially get expensive)

Yes. Inexpensive alternative: don't show the count, but _do_ show a positive indication of "empty" (checking "non-empty" for all supported services should be relatively cheap).

The goal is to avoid frustrated clicks, where user must click on things to discover that they're empty. They _already_ expanded the Region node, so why make them click on each service?

>  We need to make sure we're not locking the UI thread / stopping them from interacting with other nodes in the explorer while we do these pre-emptive fetches

Noted as an (IMPL) item.

Reply:
> That's an anxiety UX. It is a common pattern because it's easier for developers to implement, but I wouldn't elevate to the level of a convention.

I don't think it's actually because it's easier. (might not even be easier since you need a special placeholder node).

I'd say it's a deliberate UX decision because it shows the customer

* This node can have stuff under it
* It just doesn't right now because X

And probably the most importantly: it allows them to call `list` lazily if and only if the user expands the dropdown. Once you expand, you expect to see something underneath and if nothing shows up or the arrow goes away it causes confusion.

> They already expanded the Region node, so why make them click on each service?

Good reason: it lets us make all of the `list` service calls on demand.